Run services manually with go run (#1167)
neicnordic/sensitive-data-archive: DEVELOPMENT.md

# Run services with `go run`

This section explains how to run some of the services using `go run` instead of the Docker setup to facilitate development.

## Running `sda-download` with `go run`

- Bring up all SDA services with the S3 backend and populate them with test data by running the following command in the root folder of the repository:

```sh

> make integrationtest-sda-s3-run 

Reviewer: This command will build everything and run the tests and in the end will remove all containers.
So, there won't be anything running to test with.

Author: The command `make integration-sda-s3` will remove all containers when it is finished, but not this one 😄 
see https://github.com/neicnordic/sensitive-data-archive/blob/main/Makefile#L118-L119

Reviewer: Ooh, right! Should have read more carefully :blush: :

```

- Check if `sda-download` works as expected using:

```sh
curl -o /dev/null -s -w "%{http_code}\n" http://localhost:18080/health
```

If successful, the curl command should output the HTTP code `200`.

You can further check the endpoint `/metadata/datasets` using:

```sh
token=$(curl -s -k http://localhost:8080/tokens | jq -r '.[0]') 
curl -H "Authorization: Bearer $token" http://localhost:18080/metadata/datasets
```

If successful, the curl command should output a JSON body containing:

```json
["EGAD74900000101"]
```

## Running other SDA services with `go run`

Running any of the SDA services located in the `sda` subfolder requires that the service specific credentials and RabbitMQ configurations are set as ENVs. Here, we'll use `ingest` as an example.

- Bring up all SDA services with the S3 backend by running the following command in the root folder of the repository:

```sh
make sda-s3-up
```

- When the previous command is finished, bring down the `ingest` service using:

```sh
docker stop ingest
```

- Copy keys and other information from the shared folder of the container using:

```sh
docker cp verify:/shared /tmp/
```

> This will copy all data from the container's `/shared` folder to `/tmp/shared` on your local machine, so that we have access to all the required auto generated files that will be required.

Reviewer 2: ```suggestion
This will copy all data from the container's `/shared` folder to `/tmp/shared` on your local machine, so that we have access to all the auto generated files that will be required.
```

### Running `sda-download` with `go run`
1. Bring up all SDA services with the S3 backend and populate them with test data by running the following command in the root folder of the repository:
```sh
make integrationtest-sda-s3-run 
```

2. Change to the folder `sda-download` and start the `sda-download` service using:
```sh
CONFIGFILE=dev_utils/config-notls_local.yaml go run cmd/main.go
```
> If the host `oidc` cannot be accessed on the local machine, update your `/etc/hosts` file by adding the following line to ensure the service can resolve `oidc` locally:
> ```
> 127.0.0.1 oidc
> ```

3. Check if `sda-download` works as expected using:
```sh
curl -o /dev/null -s -w "%{http_code}\n" http://localhost:18080/health
```
If successful, the curl command should output the HTTP code `200`.

You can further check the endpoint `/metadata/datasets` using:
```sh
token=$(curl -s -k http://localhost:8080/tokens | jq -r '.[0]') 
curl -H "Authorization: Bearer $token" http://localhost:18080/metadata/datasets
```
If successful, the curl command should output a JSON body containing:
```json
["EGAD74900000101"]
```

> 
> ### Running other SDA services with `go run`
> Running other SDA services located in the `sda` subfolder, such as `ingest` or `verify`, differs slightly from running the `sda-download` service. Here, we'll use `ingest` as an example.
> 

Reviewer 2: ```suggestion

### Running a separate SDA service with `go run`
Running any of the SDA services located in the `sda` subfolder requires that the service specific credentials and RabbitMQ configurations are set as ENVs. Here, we'll use `ingest` as an example.

```

Author: fixed in 7f1134bb

1. Bring up all SDA services with the S3 backend and populate them with test data by running the following command in the root folder of the repository:
```sh

> make integrationtest-sda-s3-run 

Reviewer 2: ```suggestion
make sda-s3-up
```

This should probably use just the startup command that is part of PR #1162  since this will start the test script and shutting down ingest later will cause it to fail.

Author: If you just bring up the services with `sda-s3-up`, the verification test mentioned in the last point will not work. Also, if you wait until `make integrationtest-sda-s3-run` is finished, running docker stop ingest will not cause it to fail. Perhaps I should add a note to wait until the command finishes?

Author: I create a bug report https://github.com/neicnordic/sensitive-data-archive/issues/1184

Reviewer 2: It's not a bug, it's you who haven't registered the key hash in the DB before you try to ingest the file.

Author: Thanks for pointing this out. Fixed in the commit 2cefebe

```

2. Bring down the `ingest` service using:
```sh

> PR_NUMBER=$(/bin/date +%F)  docker compose -f .github/integration/sda-s3-integration.yml stop ingest  

Reviewer 2: ```suggestion
docker stop ingest  
```

You only need to call docker stop since the containers have unique names.

Author: fixed in 7f1134bb

```

3. Change to the folder `sda`, then copy keys and other information from the shared folder of the container using:
```sh
docker cp verify:/shared /tmp/
```

> This will copy all data from the container's `/shared` folder to `/tmp/shared` on your local machine.

Reviewer 2: ```suggestion
This will copy all data from the container's `/shared` folder to `/tmp/shared` on your local machine, so that we have access to all the required auto generated files that will be required.
```

Author: fixed in 7f1134bb

4. Start the `ingest` service using:
```sh
export BROKER_PASSWORD=ingest
export BROKER_USER=ingest
export BROKER_QUEUE=ingest
export BROKER_ROUTINGKEY=archived
export DB_PASSWORD=ingest
export DB_USER=ingest 
CONFIGFILE=config_local.yaml go run cmd/ingest/ingest.go
```

5. Check if the `ingest` service works as expected by following these steps
```sh
# create a test file
seq 10 > /tmp/t1.txt

# update the s3cmd config file
sed -i '/host_/s/s3inbox:8000/localhost:18000/g' /tmp/shared/s3cfg

# upload /tmp/t1.txt to s3inbox by sda-cli
sda-cli -config /tmp/shared/s3cfg upload -encrypt-with-key /tmp/shared/c4gh.pub.pem /tmp/t1.txt

# use sda-admin to check if t1.txt has been uploaded
export API_HOST=http://localhost:8090
export ACCESS_TOKEN=$(curl -s -k http://localhost:8080/tokens | jq -r '.[0]') 
sda-admin file list -user test@dummy.org # file t1.txt should have fileStatus 'uploaded'

# use sda-admin to ingest the file t1.txt
sda-admin file ingest -filepath test_dummy.org/t1.txt.c4gh -user test@dummy.org  

> # verify that t1.txt has been uploaded using sda-admin 

Reviewer 2: ```suggestion
# verify that t1.txt has been ingested using sda-admin 
```

Author: fixed in 7f1134bb

## Run services with `go run`

This section explains how to run some of the services using `go run` instead of the Docker setup to facilitate development.

### Running `sda-download` with `go run`
1. Bring up all SDA services with the S3 backend and populate them with test data by running the following command in the root folder of the repository:
```sh
make integrationtest-sda-s3-run 
```

> 
> 2. Change to the folder `sda-download` and then create the credentials using: 
> ```sh
> pushd dev_utils
> bash make_certs.sh
> chmod 600 certs/*-key.*
> popd
> ```

Reviewer 2: ```suggestion
```

Certificates are not needed since it specifically says to use the `no_tls` config file

3. Update your `/etc/hosts` file: Add the following line to ensure the service can resolve OIDC locally: 
```
127.0.0.1 oidc
```

> 

Author: Do you mean something like this in `sda-s3-integration.yml`? Doesn't work. 

```yaml
  oidc:
    container_name: oidc
    command:
      - /bin/sh
      - -c
      - |
        pip install --upgrade pip
        pip install aiohttp Authlib joserfc requests
        python -u /oidc.py
    depends_on:
      credentials:
        condition: service_completed_successfully
    healthcheck:
      test: ["CMD", "python3", "-c", 'import requests; print(requests.get(url = "http://localhost:8080/jwk").text)']
      interval: 10s
      timeout: 2s
      retries: 6
    image: python:3.11-slim
    ports:
      - "8080:8080"
    extra_hosts:
      - "oidc:host-gateway"
    restart: always
    volumes:
      - ./sda/oidc.py:/oidc.py
      - shared:/shared
```

> 

Reviewer 2: ```yaml
extra_hosts:
  - "localhost:host-gateway"
```

In the correct place will make things accessible in the same way from the developers perspective as from a container. If you are having issues, look in the GDI repo for hints on how to get it to work.

> 

Author: I think `host-gateway` can solve the problem to access a localhost from inside a container but not to access e.g. http://oidc:8080 from outside of the container, which is the case when we run `sda-download` with `go run`, right`? So this solution won't work.

> 

Reviewer 2: The whole point with the `localhost:host-gateway` mapping is to make a container callable by the same name from both the shell of the developer and from within another container.

It's possible to use another name than localhost, as long as that name exists in the `etc/hosts` file of the host machine, like for example the value of `$HOSTNAME`.

> 

Author: For some reason, the solutions for host-gateway, e.g. in this PR https://github.com/GenomicDataInfrastructure/starter-kit-storage-and-interfaces/pull/57/files, never worked on my Linux setup. Neither for Ubuntu 22.04 nor 18.04 on SNIC cloud or EGI cloud. If it works for you, cloud you please make the changes to `.github/integration/sda-s3-integration.yml`? Thanks.

> 

Reviewer 2: I can look in to it once i'm in a place to commit things for the k8s documentation.

> 

Reviewer 2: I pushed a commit that makes this work from my local shell.

> 

Author: OK. So I will keep the instruction for adding "127.0.0.1 oidc" to /etc/hosts

> 

Reviewer 2: No it should be removed, my commit fixed it.

> 

Author: fixed in 7f1134bb

4. Start the `sda-download` service using:
```sh
APP_PORT=8553 CONFIGFILE=dev_utils/config-notls_local.yaml go run cmd/main.go
```

5. Check if `sda-download` works as expected using:
```sh
token=$(curl -s -k http://localhost:8080/tokens | jq -r '.[0]') 
curl -H "Authorization: Bearer $token" http://localhost:8553/metadata/datasets
```
If successful, the curl command should output a JSON body containing:
```json
["EGAD74900000101"]
```

### Running other SDA services with `go run`
Running other SDA services located in the `sda` subfolder, such as `ingest` or `verify`, differs slightly from running the `sda-download` service. Here, we'll use `ingest` as an example.

> 1. Bring up all SDA services with the S3 backend and populate them with test data by running the following command in the root folder of the repository (this step is the same as for `sda-download`):

Reviewer 2: If you describe how to run a single command from the sda stack there is no need to reference `sda-download` and compare the differences. Keep it strictly for the sda stack.

```sh
make integrationtest-sda-s3-run 
```

2. Bring down the `ingest` service using:
```sh
PR_NUMBER=$(/bin/date +%F)  docker compose -f .github/integration/sda-s3-integration.yml stop ingest  
```

> 3. Change to the folder `sda`, then copy keys and other information from the shared folder of the container using:
> ```sh
> mkdir -p tmp
> docker cp verify:/shared tmp/
> ```

Reviewer 2: Do not copy temp files into the current workdir, always use the system temp folder.

```suggestion
3. Change to the folder `sda`, then copy keys and other information from the shared folder of the container using:
```sh
docker cp verify:/shared /tmp/shared
```
```

> This will copy all data from the container's `/shared` folder to `tmp/shared` on your local machine.

Reviewer 2: ```suggestion
This will copy all data from the container's `/shared` folder to `/tmp/shared` on your local machine.
```

> 
> 4. Copy schemas to `/schemas` on the localhost using:
> ```sh
> sudo rsync -arvz schemas/ /schemas/ 
> ```
> 

Reviewer 2: ```suggestion
```

This should not be needed, the schema path is a configurable option, although not directly so that is something we should fix to make this easier.

Author: The problem is nontrivial. Currently, the `SchemasPath` is hardcoded based on the value of `schema.type`:
https://github.com/neicnordic/sensitive-data-archive/blob/main/sda/internal/config/config.go#L919-L925

I have tried modifying this function to set `SchemasPath` to a custom path, but it still doesn't work.

We should create an issue to make `SchemasPath` customizable. In the meantime, we need to manually copy the schemas to make it work.

Reviewer 2: ```
func (c *Config) configSchemas() {
	if viper.GetString("schema.type") == "federated" {
		c.Broker.SchemasPath = "/schemas/federated/"
	} else {
		c.Broker.SchemasPath = "/schemas/isolated/"
	}

	if viper.IsSet("schema.path") {
		c.Broker.SchemasPath = viper.GetString("schema.path")
	}
}
```

Author: This works well. When I tested yesterday, I think the reason why it failed was because the MQ message are not cleaned and therefore the old schema path was used still. 
Fixed in the commits 17e9d1d2 and 73f8a274

5. Start the `ingest` service using:
```sh
export BROKER_PASSWORD=ingest
export BROKER_USER=ingest
export BROKER_QUEUE=ingest
export BROKER_ROUTINGKEY=archived
export DB_PASSWORD=ingest
export DB_USER=ingest 
CONFIGFILE=config_local.yaml go run cmd/ingest/ingest.go
```

6. Check if the `ingest` service works as expected by following these steps
```sh
# create a test file

> seq 10 > tmp/t1.txt  

Reviewer 2: ```suggestion
seq 10 > /tmp/t1.txt  
```

# update the s3cmd config file

> sed -i '/host_/s/s3inbox:8000/localhost:18000/g' tmp/shared/s3cfg

Reviewer 2: ```suggestion
sed -i '/host_/s/s3inbox:8000/localhost:18000/g' /tmp/shared/s3cfg
```

# upload tmp/t1.txt to s3inbox by sda-cli

> sda-cli -config tmp/shared/s3cfg upload -encrypt-with-key tmp/shared/c4gh.pub.pem tmp/t1.txt 

Reviewer 2: ```suggestion
sda-cli -config /tmp/shared/s3cfg upload -encrypt-with-key /tmp/shared/c4gh.pub.pem tmp/t1.txt 
```
